checks/branch_protected.go

```diff
@@ -31,47 +33,94 @@ func init() {
 	registerCheck(CheckBranchProtection, BranchProtection)
 }
 
+type repositories interface {
+	Get(context.Context, string, string) (*github.Repository,
+		*github.Response, error)
+	ListReleases(ctx context.Context, owner string, repo string, opts *github.ListOptions) (
+		[]*github.RepositoryRelease, *github.Response, error)
+	GetBranchProtection(context.Context, string, string, string) (
+		*github.Protection, *github.Response, error)
+}
+
+type logger func(s string, f ...interface{})
+
 func BranchProtection(c *checker.CheckRequest) checker.CheckResult {
-	repo, _, err := c.Client.Repositories.Get(c.Ctx, c.Owner, c.Repo)
+	// Checks branch protection on both release and development branch
+	return checkReleaseAndDevBranchProtection(c.Ctx, c.Client.Repositories, c.Logf, c.Owner, c.Repo)
+}
+
+func checkReleaseAndDevBranchProtection(ctx context.Context, r repositories, l logger, ownerStr,
+	repoStr string) checker.CheckResult {
+	// Get release branches
+	releases, _, err := r.ListReleases(ctx, ownerStr, repoStr, &github.ListOptions{})
 	if err != nil {
 		return checker.MakeRetryResult(CheckBranchProtection, err)
 	}
 
-	protection, resp, err := c.Client.Repositories.
-		GetBranchProtection(c.Ctx, c.Owner, c.Repo, *repo.DefaultBranch)
+	var checks []checker.CheckResult
+	checkedBranches := map[string]bool{}
+	for _, release := range releases {
+		// may be nil if release tag already exists
+		commitish := release.TargetCommitish
+		if commitish != nil && !checkedBranches[*commitish] {
+			res := getProtectionAndCheck(ctx, r, l, ownerStr, repoStr, *commitish)
```

Comment:
we could reduce the number of API calls by returning early if we find a non-protected branch. In most checks, though, I've taken the decision to continue anyway: this way we can show users the entire list of offending branches they need to fix. I suppose a better way would be to return early if the option `--show-details` was not provided by the user, and iterate thru all branches if the option was not provided (in a different CL).

Reply:
Actually yes. I just a nice way of resolving this and https://github.com/ossf/scorecard/pull/554#discussion_r650037369 together.

We can make one call to ListBranches at the start. This tells us which branches exist and which are protected in a single call https://github.com/google/go-github/blob/edaee9aa26ea0b2fedd961398433b8940b2afef1/github/repos.go#L912

Then from the release target_commitish and dev branch, check this list first and only check the ones with protections